[#11571] Selective Deadline Extensions - Frontend tables and modals

FergusMok opened the pull request:
Partially fixes #11571 for Milestone 1 

This PR includes
- Instructor's individual extension page for students and instructors
- Selection of extension time modal
- Confirmation modal

https://user-images.githubusercontent.com/10182564/157613720-c455a9c9-e894-4aa9-994b-d3a32795b560.mp4

Current behavior of extension:
1. You should be able to select individuals with and without extensions to update / create extension at the same time.
2. You should only be able to select individuals with pre-existing extensions to delete.

To be handled in a separate PR:
1. Toggleable tables for the instructors / students
2. Status of extensions for the feedback session from the feedback session page

moziliar commented:
Would be good to also include a snapshot of the UI in the PR.

moziliar commented:
@FergusMok Any updates on this PR?

FergusMok commented:
@moziliar Yup, sorry for the delay. I updated the PR with a short video on the UI

samuelfangjw commented:
@FergusMok Seems like `package-lock.json` is modified, could you help check if this is intended?

FergusMok commented:
@samuelfangjw Yup thanks for pointing that out. It was unintentional

moziliar commented:
The UI seems to honour the tech design!

Some comments I have about the UI:

* Will the dropdown bar look nicer if it is shorter?
* For the text below the dropdown: should it be "The deadline will be extended from T1 to T2" so that the user can refer to the old one quicker?
* Will the confirmation page with the students affected be included too?

Can also look into how E2E can be done.

samuelfangjw commented:
@FergusMok, can help to rebase only the relevant changes only selective deadline ext branch? Shouldn't include irrelevant commits from master.

FergusMok commented:
>

FergusMok commented:
@moziliar Hi Zongran

> Some comments I have about the UI:
> 
>     * Will the dropdown bar look nicer if it is shorter?

Are you talking about the options for "Extend By"?

>     * For the text below the dropdown: should it be "The deadline will be extended from T1 to T2" so that the user can refer to the old one quicker?

This makes sense, but it'll be impossible to get T1 if we are trying to extend multiple individuals with pre-existing extensions

>     * Will the confirmation page with the students affected be included too?
Yup, the confirmation page is the 2nd modal

samuelfangjw commented on src/main/java/teammates/ui/request/DeadlineExtensionRequest.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * The request specifying deadline extensions to create, update or revoke.
+ */
+public class DeadlineExtensionRequest extends BasicRequest {
+
+    private String courseId;
+    private String feedbackSessionName;
+    private boolean notifyUsers;
+    private Map<String, Long> studentExtensionsToModify;
+    private List<String> studentExtensionsToRevoke;
+    private Map<String, Long> instructorExtensionsToModify;
+    private List<String> instructorExtensionsToRevoke;
```
This seems like it came from my PR, why is this necessary?

FergusMok replied:
My bad, I was picking out the requests files to generate the types. Will remove

FergusMok commented:
@samuelfangjw Yup thanks for the catch, I just rebased

moziliar commented:
> Are you talking about the options for "Extend By"?

Yep. It looks unnecessarily long?

> This makes sense, but it'll be impossible to get T1 if we are trying to extend multiple individuals with pre-existing extensions

I see. In that case, just leave it to the confirmation page.

> Yup, the confirmation page is the 2nd modal

Got it! Changes look good!

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+  onConfirm(): void {
+    if (this.getExtensionTimestamp() < Date.now()) {
+      const extensionTimeString = this.timeZoneService
+      .formatToString(this.getExtensionTimestamp(), this.feedbackSessionTimeZone, this.DATETIME_FORMAT);
+      const currentTimeString = this.timeZoneService
+      .formatToString(Date.now(), this.feedbackSessionTimeZone, this.DATETIME_FORMAT);
```
Should we add some indentation here to distinguish?

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+      .formatToString(Date.now(), this.feedbackSessionTimeZone, this.DATETIME_FORMAT);
+      this.simpleModalService.openConfirmationModal(
+        'Are you sure you want to extend to before the current time?',
+        SimpleModalType.DANGER,
```
@damithc Is `DANGER` better, or should we use `WARNING` instead?

damithc replied:
Probably WARNING is a better match? We usually use DANGER for things such as deleting.
I didn't get the meaning of the `extend to before the current time` phrase though.

moziliar replied:
How about `"Are you sure you wish to set the new deadline to before the current time?"`? Can also add in consequences such as "`Any users affected will have their session closed immediately.`". Though, it is hard to tell what the real consequence is should we allow such behaviour, as it should rightfully invalidate any submissions after the updated deadline even if it is submitted seconds before this action.

damithc replied:
Perhaps just disallow it? Extensions by definition should be later than the normal deadline. I don't think anybody will ask for a feature that allows the opposite of extensions.

FergusMok replied:
Sorry for the vague phrasing. The modal is for the case when there's a pre-existing extension, but the instructor is (accidentally) updating to a time inbetween the original deadline and the current exact time (reducing the extension), effectively closing the session to those who have the extensions. 

The case of opposite extensions is already disallowed (cannot pick with the timepicker) 

I agree with @moziliar's suggestions for the improvement to make it clearer. 

For the mentioned behavior, then would the undefined behavior extend to deletion too? Im thinking then we'll need a warning to instructors to not delete / reduce extensions to those who's submissions will be invalidated by the action. Although this will add additional complexity (most likely in the frontend's validation)

FergusMok replied:
Moving the conversation about undefined behavior to the backend PR: https://github.com/TEAMMATES/teammates/pull/11649#issuecomment-1073057610

FergusMok replied:
@jayasting98 @samuelfangjw Just adding you guys in for visibility. Also adding in this discussion https://github.com/TEAMMATES/teammates/pull/11649#issuecomment-1073057610. 

The frontend now should have additional validation for:
1. Moving the sessionEndTime to before any extensions (in the sessions edit page), we'll warn of the invalidation of some extensions.
2. Deleting Extensions / Updating pre-existing extensions to before the current date (in the extension page), we'll warn of immediate closure of the session for affected individuals.

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  onChangeDateTime(data: DateFormat | TimeFormat, field: string): void {
+    if (field === 'date') {
+      this.datePicker = data as DateFormat;
+    }
+    if (field === 'time') {
+      this.timePicker = data as TimeFormat;
+    }
```
Should we use something like `if-else if-else` instead of these fallthrough looking blocks. `if-else if` shows the dichotomy better I feel:)

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+    }
+    if (this.isRadioExtendTo()) {
+      const timestamp = this.timeZoneService
+      .resolveLocalDateTime(this.datePicker, this.timePicker, this.feedbackSessionTimeZone, true);
```
ditto

moziliar reviewed (changes requested):
Thanks for the hard work! Just checking if there's any reason to make the file `src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.scss`?

halfwhole commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,512 @@
+          this.instructorDeadlines = updatedFeedbackSession.instructorDeadlines ?? {};
+          this.statusMessageService.showSuccessToast(
+          `Extension for ${studentsSelected.length} students and ${instructorsSelected.length} instructors
+           have been successful!`);
```
Maybe extension [...] has been, or extensions [...] have been successful?

FergusMok replied:
Yup, thanks for the catch

halfwhole commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,203 @@
+        <div class="row text-center">
+          <div class="col-md-2 text-md-right font-bold">Original Deadline</div>
+          <div id="course-name" class="col-md-10 text-md-left">
+            {{ timezoneService.formatToString(feedbackSessionEndingTime, feedbackSessionTimeZone, DATETIME_FORMAT) }}
```
It could be better to perform this formatting in the logic instead, and only after we've fetched the feedback session information - if the arguments are still their default values (before we've fetched the info), errors can show up

<img width="581" alt="Screenshot 2022-03-20 at 3 16 35 PM" src="https://user-images.githubusercontent.com/41856541/159152372-e5080164-81f0-4cd8-a5bf-4998046c3a55.png">

FergusMok replied:
The problem was that timezone default's was ''. I've put the default as 'UTC'. Because loading only stops after the correct zone is fetched, should not be a problem

halfwhole reviewed (changes requested):
UI-wise it looks pretty good already! Still have a number of comments though, with regard to some of the logic and the variable namings:

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.html:
```diff
@@ -0,0 +1,88 @@
+        <label for="form-minutes">Minutes</label>
+        <input class="form-control mb-2" name="form-minutes" type="text" placeholder="Minutes"
+        [(ngModel)]="extendByDatePicker.minutes">
+      </div>
```
In hindsight, I agree, minutes may not be useful (Picking the feedback session time is in hourly chunks anway)

FergusMok commented on src/main/java/teammates/ui/request/FeedbackSessionUpdateRequest.java:
```diff
@@ -1,8 +1,54 @@
+    private Map<String, Long> studentDeadlines;
+    private Map<String, Long> instructorDeadlines;
+
+    private boolean isGoingToNotifyAboutDeadlines;
```
This file was added from another PR to generate types for the frontend. Tagging @jayasting98 for visibility

samuelfangjw replied:
Just to keep everyone in the loop, after discussion with jay we felt it's best to move this out of the request body and send this as a query param instead. This will be removed in his PR!

https://github.com/TEAMMATES/teammates/pull/11649#discussion_r830504851

jayasting98 replied:
>Seems a bit lengthy

Yea I was trying to make it start with "is". Samuel has told me to shorten it too, which I will do.

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,512 @@
+          }))
+          .subscribe(() => {
+              this.getAllStudentsOfCourse(); // Both students and instructors need feedback ending time.
+              this.getAllInstructorsOfCourse();
```
The problem currently is that we need the feedback session end time, before we can render the 'current deadline' field for the individuals. So there's actually a dependency here

halfwhole replied:
Ah good point! My mistake here

FergusMok commented on src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts:
```diff
@@ -433,6 +433,11 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
+      studentDeadlines: {},
+      instructorDeadlines: {},
+
+      isGoingToNotifyAboutDeadlines: false,
```
I've decided to just plug it back as-is, as they'll be retrieved either way under the session

FergusMok replied:
Actually I realized it may actually be better to make the fields optional. I'll liaise with Jay about it.

samuelfangjw replied:
> Actually I realized it may actually be better to make the fields optional. I'll liaise with Jay about it.

As discussed on slack earlier, ideally we want to try to keep this here because it is a` POST` and not a `PATCH` request. Will also simplify updating logic.

FergusMok replied:
Got it, thanks!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.ts:
```diff
@@ -0,0 +1,148 @@
+
+  DATETIME_FORMAT: string = 'd MMM YYYY h:mm:ss';
+
+  isNotifyStudents: boolean = false;
```
Yup, thanks for the catch

FergusMok commented on src/web/app/components/session-edit-form/session-edit-form.component.ts:
```diff
@@ -20,6 +21,7 @@ import { SessionEditFormMode, SessionEditFormModel } from './session-edit-form-m
 /**
  * Form to Add/Edit feedback sessions.
  */
+
```
Nope, I think this was accidental! Same for the few points below. Will remove them

FergusMok commented on src/web/services/datetime-helper.ts:
```diff
@@ -0,0 +1,16 @@
+@Injectable()
+export class DateTimeHelper {
+    static addTime(timestamp: number, minutes: number, hours: number, days: number): number {
+        return timestamp + (minutes * MINUTES_TO_MILLI) + (hours * HOURS_TO_MILLI) + (days * DAY_TO_MILLI);
```
I've deleted the file, was originally planning to refactor. Will handle in separate PR!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,512 @@
+        this.statusMessageService.showErrorToast(resp.error.message);
+        this.hasLoadingFeedbackSessionFailed = true;
+        this.isLoadingAllStudents = false;
+        this.isLoadingAllInstructors = false;
```
Solved. Currently only instructor does not throw an error with an invalid course id (Might be a backend problem)

FergusMok commented:
I removed the institute field from the Instructor table, and replaced it with Roles. I thought it'll be redundant, since in most (if not all) cases, the instructors are coming from the same institute.

samuelfangjw commented:
> I removed the institute field from the Instructor table, and replaced it with Roles. I thought it'll be redundant, since in most (if not all) cases, the instructors are coming from the same institute.

Yes, probably a good idea. Institute is now (or will very soon) be a property of the course and not the instructor account. The institute field in accounts shouldn't be used.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.html:
```diff
@@ -0,0 +1,161 @@
+<div class="modal-body" >
+  <div *ngIf="isExtendModal()">
+    The new deadline of <b>{{ extensionTimestamp | formatDateDetail: feedbackSessionTimeZone }}</b> will be applied
+    to the following <b>{{selectedStudents.length}}</b> student(s) and <b>{{selectedInstructors.length}}</b> instructor(s) upon confirmation:
```
I don't think we need the words "on confirmation"? I don't think this adds much value to the sentence.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+          strB = b.email;
+          break;
+        case SortBy.INSTRUCTOR_PERMISSION_ROLE:
+          strA = a.role ? a.role.valueOf() : '';
```
To check: `valueOf()` may not be necessary. The corresponding method in `InstructorSessionIndividualExtensionPageComponent` didn't have this.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.ts:
```diff
@@ -0,0 +1,143 @@
+  onConfirmExtensionCallBack: EventEmitter<boolean> = new EventEmitter();
+
+  constructor(public activeModal: NgbActiveModal,
+              public timezoneService: TimezoneService,
```
is timezone service necessary?

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.html:
```diff
@@ -0,0 +1,84 @@
+  </button>
+</div>
+<div class="modal-body">
+  Extending the feedback session for <b>{{numStudents}}</b> student(s) and <b>{{numInstructors}}</b> instructor(s):
```
nit: To make it consistent with the other variables in this file
```suggestion
  Extending the feedback session for <b>{{ numStudents }}</b> student(s) and <b>{{ numInstructors }}</b> instructor(s):
```

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,147 @@
+
+  getExtensionTimestamp(): number {
+    if (this.isRadioExtendBy()) {
+      if (!this.isCustomize() && this.extendByDeadlineOptions.has(this.extendByDeadlineKey)) {
+        return this.addTime(this.feedbackSessionEndingTime,
+          this.extendByDeadlineOptions.get(this.extendByDeadlineKey)!.valueOf(), 0);
+      }
+      if (this.isCustomize()) {
+        return this.addTime(this.feedbackSessionEndingTime, this.extendByDatePicker.hours,
+          this.extendByDatePicker.days);
+      }
```
```suggestion
      if (this.isCustomize()) {
        return this.addTime(this.feedbackSessionEndingTime, this.extendByDatePicker.hours,
          this.extendByDatePicker.days);
      } else if (this.extendByDeadlineOptions.has(this.extendByDeadlineKey)) {
        return this.addTime(this.feedbackSessionEndingTime,
          this.extendByDeadlineOptions.get(this.extendByDeadlineKey)!.valueOf(), 0);
      }
```

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.spec.ts:
```diff
@@ -0,0 +1,115 @@
+        expect(component).toBeTruthy();
+      });
+
+    it('should snap with default fields', () => {
+        expect(component).toBeTruthy();
+    });
```
Test does not match description, this should be a snapshot test with default values.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.spec.ts:
```diff
@@ -0,0 +1,432 @@
+        expect(component).toBeTruthy();
+      });
+
+    it('should snap with default fields', () => {
+        expect(component).toBeTruthy();
+    });
```
Test does not match description, this should be a snapshot test with default values.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.spec.ts:
```diff
@@ -0,0 +1,103 @@
+        expect(component).toBeTruthy();
+    });
+
+    it('should snap with default fields', () => {
+        expect(component).toBeTruthy();
+    });
```
Test does not match description, this should be a snapshot test with default values.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,201 @@
+      </div>
+    </div>
+    <div class="container ml-1">
+      <p> Select students or instructors from list below and click apply or remove to change their feedback session deadline
```
The names of the buttons need to be updated as well.

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,517 @@
+  }
+
+  constructor(
+    public timezoneService: TimezoneService,
```
Nope, it was from an older version when I wasn't using pipes yet

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.spec.ts:
```diff
@@ -0,0 +1,432 @@
+        feedbackSessionsService = TestBed.inject(FeedbackSessionsService);
+        timezoneService = TestBed.inject(TimezoneService);
+        statusMessageService = TestBed.inject(StatusMessageService);
+        jest.spyOn(timezoneService, 'guessTimezone').mockReturnValue('UTC');
```
Again, from an older version. Thanks

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,517 @@
+      );
+  }
+
+  private getFeedbackSessionDetails(feedbackSession: FeedbackSession): void {
+    this.feedbackSessionDetails = {
+      instructions: feedbackSession.instructions,
+      submissionStartTimestamp: feedbackSession.submissionStartTimestamp,
+      submissionEndTimestamp: feedbackSession.submissionEndTimestamp,
+      gracePeriod: feedbackSession.gracePeriod,
+      sessionVisibleSetting: feedbackSession.sessionVisibleSetting,
+      customSessionVisibleTimestamp: feedbackSession.customSessionVisibleTimestamp,
+      responseVisibleSetting: feedbackSession.responseVisibleSetting,
+      customResponseVisibleTimestamp: feedbackSession.customResponseVisibleTimestamp,
+      isClosingEmailEnabled: feedbackSession.isClosingEmailEnabled,
+      isPublishedEmailEnabled: feedbackSession.isPublishedEmailEnabled,
+    };
+    this.feedbackSessionEndingTime = feedbackSession.submissionEndTimestamp;
+    this.feedbackSessionTimeZone = feedbackSession.timeZone;
+    this.studentDeadlines = feedbackSession.studentDeadlines ?? {};
+    this.instructorDeadlines = feedbackSession.instructorDeadlines ?? {};
+  }
```
Good idea, renamed to `setFeedbackSessionDetails`

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.module.ts:
```diff
@@ -0,0 +1,50 @@
+import { InstructorSessionIndividualExtensionPageComponent }
+from './instructor-session-individual-extension-page.component';
+
+const routes: Routes = [
+    {
+      path: '',
+      component: InstructorSessionIndividualExtensionPageComponent,
+    },
+  ];
```
Yup, my bad. I haven't configured my formatter when I wrote this.

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,517 @@
+    this.hasLoadingFeedbackSessionFailed = false;
+    this.isLoadingAllInstructors = true;
+    this.hasLoadedAllStudentsFailed = false;
+    this.courseService
+      .getCourseAsInstructor(this.courseId)
+      .pipe(finalize(() => { this.isLoadingFeedbackSession = false; }))
+      .subscribe((course: Course) => {
+        this.courseName = course.courseName;
+        this.feedbackSessionsService
+          .getFeedbackSession({
+            courseId: this.courseId,
+            feedbackSessionName: this.feedbackSessionName,
+            intent: Intent.FULL_DETAIL,
+          })
```
Yup, you're right. They should be done within a forkjoin

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,201 @@
+<tm-loading-retry [shouldShowRetry]="hasLoadingFeedbackSessionFailed" [message]="'Failed to load course and feedback session'" (retryEvent)="loadFeedbackSessionAndIndividuals()">
+  <div *tmIsLoading="isLoadingFeedbackSession">
+    <div class="card border-primary margin-top-20px">
+      <div class="card-body">
+        <div class="row text-center">
```
![image](https://user-images.githubusercontent.com/10182564/160276355-ceb77fe7-64f8-4be9-9973-daaebbb24d16.png)

Something like this?

samuelfangjw replied:
Yes, I personally think that looks much nicer!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,201 @@
+      <p> Select students or instructors from list below and click apply or remove to change their feedback session deadline
+        extension. </p>
+      <div class="align-row">
+        <div>
+          <button type="button" id="extend-btn" class="btn btn-success" (click)="onExtend()" [disabled]="!hasSelected()">Extend / Edit</button>
+        </div>
+        <div>
+          <button type="button" id="delete-btn" class="btn btn-danger" (click)="onDelete()" [disabled]="!hasSelectedValidForDeletion()">Delete</button>
+        </div>
```
That is true, it makes more intuitive sense to see the deadline as a `form`, with the buttons at the bottom. I'll make the changes

FergusMok replied:
I made the changes, and I realized it's not too intuitive.

For other parts of the application, you have to press a Edit/Delete button before proceeding to do the editing.
Without seeing the blanked out buttons, some people may be trying to find the button to enter the editing 'mode', and may not choose the individuals first. 

Although this change is not 100% intuitive for now, with the toggleable tables, the user should still be able to see the buttons before expanding the tables, so it'll probably be more intuitive then.

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.scss:
```diff
@@ -0,0 +1,54 @@
+
+.tst {
+  width: 1000px;
+}
```
Deleted! It's not used

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.scss:
```diff
@@ -0,0 +1,54 @@
+
+.remind-all {
+  width: 250px;
+}
```
Same as above, unused and deleted

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.scss:
```diff
@@ -0,0 +1,3 @@
+.hours {
+  padding-left: 0;
+}
```
Seems like this was deleted, let's leave the empty file in. The other pages without custom styles have the empty scss file as well.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,503 @@
+  private getAllStudentsOfCourse(): void {
+    this.studentService
+      .getStudentsFromCourse({ courseId: this.courseId })
+      .pipe(finalize(() => { this.isLoadingAllStudents = false; }),
+        map((students: Students) => {
+          this.studentsOfCourse = students.students.map((student) => this.mapStudentToStudentModel(student));
+        }))
+      .subscribe(() => {
+        this.initialSortOfStudents();
+      },
```
As far as possible, try to avoid side effects within pipe. If you have to perform operations with side effects, use `tap` instead of map. In this case, something like this would be suffice.

```suggestion
      .pipe(finalize(() => { this.isLoadingAllStudents = false; }),
        map((students: Students) => students.students.map((student) => this.mapStudentToStudentModel(student))),
      )
      .subscribe(
        (students: StudentExtensionTableColumnModel[]) => {
          this.studentsOfCourse = students;
          this.initialSortOfStudents();
        },
```

samuelfangjw replied:
The same applies for the corresponding method for instructors

FergusMok replied:
Makes sense, good catch, I didn't give the side effects portion much thought before

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,503 @@
+    .pipe(finalize(() => { this.isSubmittingDeadlines = false; }))
+    .subscribe(() => {
+      this.loadFeedbackSessionAndIndividuals();
+      this.statusMessageService.showSuccessToast(
+      `Deletion of extension(s) for ${selectedStudents.length} student(s) and ${selectedInstructors.length} `
+      + 'instructor(s) have been successful!');
```
This probably should be indented

moziliar replied:
Recommend to just configure the IDE/Editor you are using to auto fix it:)

samuelfangjw reviewed (changes requested):
Thanks for the changes, generally seems fine to me, should be able to merge soon!

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -152,16 +148,14 @@ export class InstructorSessionIndividualExtensionPageComponent implements OnInit
         map((students: Students) => {
-          this.studentsOfCourse = students.students.map((student) => this.mapStudentToStudentModel(student));
+          return students.students.map((student) => this.mapStudentToStudentModel(student));
         }),
```
Can do without the return here, just `(students: Students) => students.students.map...`

If you want to get rid of the awkward `students.students` you can even add another map.

```
map((students:Students) => students.students),
map((students:Student[]) => students.mapStudentToStudentModel),
```

FergusMok replied:
Thanks, I just did `({ students }: Students) => students.map`

FergusMok replied:
Actually, I refactored it. So the `mapStudentToStudentModel` is now `mapStudentsToStudentModels`. It's cleaner to look at now

samuelfangjw commented:
Overall changes look good! I'll take a closer look tomorrow but I doubt I'll have many more comments.

samuelfangjw commented:
@FergusMok, I got this while testing on localhost (the date is not 1 march). I'll look into this, but letting u know in case you know what the issue is.

<img width="1552" alt="Screen Shot 2022-03-28 at 7 41 14 PM" src="https://user-images.githubusercontent.com/60355570/160390570-1ea377b0-64eb-45c8-b645-e81fe123a42e.png">

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,539 @@
+        () => {
+          this.loadFeedbackSessionAndIndividuals();
+          this.statusMessageService.showSuccessToast(
+            `Deletion of extension(s) for ${selectedStudents.length} student(s) and ${selectedInstructors.length} `
+              + 'instructor(s) have been successful!',
```
Have been successful seems grammatically wrong here. 

Maybe 
```suggestion
            `Successfully deleted extension(s) for ${selectedStudents.length} student(s) and ${selectedInstructors.length} `
              + 'instructor(s)',
```

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,539 @@
+      .pipe(finalize(() => { this.isSubmittingDeadlines = false; }))
+      .subscribe(() => {
+          this.loadFeedbackSessionAndIndividuals();
+          this.statusMessageService.showSuccessToast(
+            `Extension(s) for ${selectedStudents.length} student(s) and ${selectedInstructors.length} instructor(s)
+           have been successful!`,
+          );
```
Same here

samuelfangjw commented:
One more issue:

Select All Students -> Extend/Delete -> ... -> Confirm -> Select all still selected, but nothing is selected.

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,155 @@
+    ['Customize', 0],
+  ]);
+
+  DATETIME_FORMAT: string = 'd MMM YYYY h:mm:ss';
```
Actually this format is unused. I'll delete it.
The actual format being used is `ddd, DD MMM YYYY, hh:mm A z` via a pipe. The difference is day and timezone. I feel both fields may be important to a instructor. I may want confirmation on the timezone of the extensions, and also the day would be useful. A portion of extensions would usually be relative to a certain day in the week (e.g extension till Saturday)

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,208 @@
+        </div>
+      </div>
+    </div>
+    <div class="container ml-1">
```
One last minor UI nit, 
<img width="1357" alt="Screen Shot 2022-03-30 at 8 05 15 PM" src="https://user-images.githubusercontent.com/60355570/160830594-610dc440-aafa-4792-a93c-3fca99c65baa.png">

Can remove the margin around the student/instructor tables and the text? Unless this is intentional, but I doubt so.

```suggestion
    <div>
```

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,208 @@
+<tm-loading-retry [shouldShowRetry]="hasLoadedAllStudentsFailed" [message]="'Failed to load students'"
+  (retryEvent)="loadFeedbackSessionAndIndividuals()">
+  <div *tmIsLoading="isLoadingAllStudents">
+    <div class="modal-body">
```
```suggestion
    <div>
```

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,208 @@
+<tm-loading-retry [shouldShowRetry]="hasLoadedAllInstructorsFailed" [message]="'Failed to load instructors'"
+  (retryEvent)="loadFeedbackSessionAndIndividuals()">
+  <div *tmIsLoading="isLoadingAllInstructors">
+    <div class="modal-body">
```
```suggestion
    <div>
```

samuelfangjw reviewed (changes requested):
One last nit. @moziliar, @halfwhole do y'all have anything to add?

halfwhole commented on src/web/services/feedback-sessions.service.ts:
```diff
@@ -103,12 +103,13 @@ export class FeedbackSessionsService {
       courseid: courseId,
       fsname: feedbackSessionName,
-      notifydeadlines: 'false',
+      notifydeadlines: isNotifyDeadlines ? 'true' : 'false',
```
nit: this works well, but would a `toString` be possible?

halfwhole commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,544 @@
+    this.hasLoadingFeedbackSessionFailed = false;
+    this.isLoadingAllInstructors = true;
+    this.hasLoadedAllStudentsFailed = false;
+    forkJoin([
+      this.courseService.getCourseAsInstructor(this.courseId),
+      this.feedbackSessionsService.getFeedbackSession({
+        courseId: this.courseId,
+        feedbackSessionName: this.feedbackSessionName,
+        intent: Intent.FULL_DETAIL,
+      }),
+    ])
```
just a sidenote! I'm good with this as it is but it could lead to indeterminacy in error messages - for e.g. if a course ID is incorrect, the error message could either be from getting the course or the feedback session

FergusMok replied:
That is true too. I believe for the majority of use cases, this should be provide a better user experience, so I'll leave it here for now

moziliar replied:
Try not to use forkJoin which will create burst of network calls. Only use it when you can justify why it must be added:)

halfwhole replied:
Would it be ok if there's just 2 calls in this case? though yeah not sure if using forkJoin here would be that idiomatic - I'm kinda alright with either approach (also for ref https://github.com/TEAMMATES/teammates/pull/11603#discussion_r835206407)

samuelfangjw replied:
Personally don't see any issues with this, considering forkJoin is used for many similar calls in the frontend.

moziliar replied:
It is ok then. The thing about forkJoin is that we will most likely have 2 servlet for the same request (instead of 1) and the Google API gateway would spin up extra instances if there are too many concurrent requests. I personally think the use of it should be justified (a way better UX).

halfwhole reviewed (commented):
Looks great! Only have a few minor things left to discuss:

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,544 @@
+
+    modalRef.componentInstance.onConfirmExtensionCallBack.subscribe((isNotifyDeadlines: boolean) => {
+      this.handleCreateDeadlines(selectedStudents, selectedInstructors, extensionTimestamp, isNotifyDeadlines);
+      modalRef.close();
```
Yup, I missed this out. Will add this in

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.html:
```diff
@@ -0,0 +1,82 @@
+      <div class="row">
+        <div class="col-sm pl-0">
+          <small>Hours</small>
+          <input class="form-control" id="form-hours" name="form-hours" type="text" placeholder="Hours"
+            [(ngModel)]="extendByDatePicker.hours">
+        </div>
+        <div class="col-sm">
+          <small>Days</small>
+          <input class="form-control" id="form-days" name="form-days" type="text" placeholder="Days"
```
From what I've found, it's impossible with just HTML alone. I've added 3 types of validation, and a error message in the prompt. Would that be fine?

halfwhole replied:
yup sounds good!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.html:
```diff
@@ -0,0 +1,82 @@
+    </form>
+    <br>
+    <p class="mb-0">
+      The new extended deadline will be: {{ addTimeAndFormat(extendByDatePicker.hours, extendByDatePicker.days) }}
```
Solved with the below validations

FergusMok commented:
Small change: I've made it such that clicking any part of the row selects the individual

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.html:
```diff
@@ -0,0 +1,204 @@
+  </div>
+</tm-loading-retry>
+
+<!-- TODO: Use lazy loading toggleable dropdowns-->
```
Yup, PR for it is opened in #11692

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.html:
```diff
@@ -0,0 +1,85 @@
+      </div>
+    </form>
+    <br>
+    <p *ngIf="!isCustomizeDateTimeIntegers()" class="red-text">Please input only integers into both blanks.</p>
+    <p *ngIf="!isDateSelectedLaterThanCurrentEndingTimestamp()" class="red-text">Please input a date later than the current session ending date.</p>
+    <p *ngIf="!isCustomizeBeforeMaxDate()" class="red-text">Please input a valid date.</p>
```
Yup, looks much better with the cross

halfwhole reviewed (approved):
LGTM, thanks a lot for your work!

samuelfangjw commented:
Rebased. Need to remember to remove the CI stuff before merging!

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+    return record;
+  }
+
+  private handleDeleteDeadlines(
+    selectedStudents: StudentExtensionTableColumnModel[],
+    selectedInstructors: InstructorExtensionTableColumnModel[],
+    isNotifyDeadlines: boolean,
+  ): void {
+    const request: FeedbackSessionUpdateRequest = {
+      studentDeadlines: this.getDeletedDeadlines(selectedStudents, true),
+      instructorDeadlines: this.getDeletedDeadlines(selectedInstructors, false),
+      ...this.feedbackSessionDetails,
+    };
+
+    this.isSubmittingDeadlines = true;
+    this.feedbackSessionsService
+      .updateFeedbackSession(this.courseId, this.feedbackSessionName, request, isNotifyDeadlines)
+      .pipe(finalize(() => { this.isSubmittingDeadlines = false; }))
+      .subscribe(
+        () => {
+          this.loadFeedbackSessionAndIndividuals();
+          this.statusMessageService.showSuccessToast(
+            `Successfully deleleted extension(s) for ${selectedStudents.length} student(s) and`
+            + ` ${selectedInstructors.length} instructor(s)!`,
+          );
+        },
+        (resp: ErrorMessageOutput) => {
+          this.statusMessageService.showErrorToast(resp.error.message);
+        },
+      );
+  }
```
Looks almost the same as `handleCreateDeadlines`. Can we abstract it into a handler that takes in the success toast as  a text or function?

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+      );
+  }
+
+  private getDeletedDeadlines(
+    selectedIndividuals: StudentExtensionTableColumnModel[] | InstructorExtensionTableColumnModel[],
+    isStudent: boolean,
+  ): Record<string, number> {
+    let record: Record<string, number> = {};
+    if (isStudent) {
+      record = { ...this.studentDeadlines };
+    } else {
+      record = { ...this.instructorDeadlines };
+    }
+
+    selectedIndividuals.forEach((x) => {
+      delete record[x.email];
+    });
+    return record;
+  }
```
Likewise, this is highly duplicated with `getUpdatedDeadlines`. Can we have some abstraction that allow us to reduce the duplicated lines?

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+  }
+
+  hasSelected(): boolean {
+    for (const student of this.studentsOfCourse) {
+      if (student.isSelected) return true;
+    }
+    for (const instructor of this.instructorsOfCourse) {
+      if (instructor.isSelected) return true;
+    }
```
Hack, but would this be simpler if we do the following:
```suggestion
    [...this.studentsOfCourse, ...this.instructorsOfCourse]
      .some(user => user.isSelected);
```

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+  /**
+   * Checks if at least one valid extension has been selected
+   */
+  hasSelectedValidForDeletion(): boolean {
```
ditto

moziliar reviewed (changes requested):
Great job! Some comments on the code quality:)

samuelfangjw commented:
@FergusMok, seems like there still some issue with the time. Can you investigate? On a side note, we probably need a couple of tests for this (if not now then in the future) to ensure the timings are accurate.

<img width="759" alt="Screen Shot 2022-04-04 at 8 57 43 PM" src="https://user-images.githubusercontent.com/60355570/161548826-90960c73-98e1-45c0-9f9b-de75245bfe86.png">

FergusMok commented:
@samuelfangjw Thanks for the catch, I used 0.5 hours instead of 0.5 days. Fixed

samuelfangjw commented:
> @samuelfangjw Thanks for the catch, I used 0.5 hours instead of 0.5 days. Fixed, will add some tests

Thanks for the quick fix!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+  }
+
+  private setFeedbackSessionDetails(feedbackSession: FeedbackSession): void {
+    this.feedbackSessionDetails = {
+      instructions: feedbackSession.instructions,
+      submissionStartTimestamp: feedbackSession.submissionStartTimestamp,
+      submissionEndTimestamp: feedbackSession.submissionEndTimestamp,
+      gracePeriod: feedbackSession.gracePeriod,
+      sessionVisibleSetting: feedbackSession.sessionVisibleSetting,
+      customSessionVisibleTimestamp: feedbackSession.customSessionVisibleTimestamp,
+      responseVisibleSetting: feedbackSession.responseVisibleSetting,
+      customResponseVisibleTimestamp: feedbackSession.customResponseVisibleTimestamp,
+      isClosingEmailEnabled: feedbackSession.isClosingEmailEnabled,
+      isPublishedEmailEnabled: feedbackSession.isPublishedEmailEnabled,
+    };
```
![image](https://user-images.githubusercontent.com/10182564/161705734-ae3ad150-cf33-4565-9383-be2a78cd1f9f.png)

Apparently not ☹️

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,569 @@
+  /**
+   * Handles the opening the confirmation modal to delete deadlines.
+   */
+  onDelete(): void {
```
onDelete is used in the html file so it can't be private (while `onConfirmExtension` is not)

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,181 @@
+    return this.feedbackSessionEndingTimestamp;
+  }
+
+  addTimeAndFormat(hours: number, days: number): string {
+    const time = this.addTime(this.feedbackSessionEndingTimestamp, hours, days);
+    return this.dateDetailPipe.transform(time, this.feedbackSessionTimeZone);
+  }
```
Would `addToAndFormatEndTime` be better? I was trying to convey that it adds time and formats the result

moziliar replied:
`extendAndFormatEndTimeBy(hours: number, days: number)` is probably more ergonomic. What do you think?

FergusMok replied:
Sounds better

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -123,23 +123,25 @@ export class InstructorSessionIndividualExtensionPageComponent implements OnInit
-          this.isLoadingAllInstructors = false;
-        },
-      );
+    .pipe(finalize(() => {
+      this.isLoadingFeedbackSession = false;
+      this.isLoadingAllStudents = false;
+      this.isLoadingAllInstructors = false;
+    }))
+    .subscribe(
+      (value: any[]) => {
+        const course = value[0] as Course;
+        this.courseName = course.courseName;
+        const feedbackSession = value[1] as FeedbackSession;
+        this.setFeedbackSessionDetails(feedbackSession);
+        this.getAllStudentsOfCourse(); // Both students and instructors need feedback ending time.
+        this.getAllInstructorsOfCourse();
+      },
+      (resp: ErrorMessageOutput) => {
+        this.statusMessageService.showErrorToast(resp.error.message);
+        this.hasLoadingFeedbackSessionFailed = true;
+      },
+    );
```
Indentation is gone?

FergusMok replied:
My bad, I mistakenly unindented it

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -154,21 +154,21 @@ export class IndividualExtensionDateModalComponent {
-      if (!Number.isInteger(this.extendByDatePicker.days) || !Number.isInteger(this.extendByDatePicker.hours)) {
-        return false;
-      }
+    if (!this.isCustomize()) {
```
`isCustomize` => `isCustomized`?

moziliar replied:
I agree, though it actually means is "customize"... well can't think of a more appropriate name for now

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.scss:
```diff
@@ -0,0 +1,3 @@
+.text-danger {
+  margin: 0;
+}
```
Let's add `m-0` to the class instead. I don't think we should be overriding the bootstrap css classes unless necessary.

FergusMok replied:
![image](https://user-images.githubusercontent.com/10182564/161887465-2d97a103-b296-4f78-9eae-d92f04ca2b01.png)

Bootstrap's implementation of m-0 is the same, would there be a difference?

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,189 @@
+    return this.isRadioExtendBy() && this.extendByDeadlineKey === 'Customize';
+  }
+
+  sortMapByOriginalOrder = (): number => {
+    return 0;
+  };
```
This is a field, better to place this above with the rest.

samuelfangjw commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,527 @@
+      );
+  }
+
+  private showSucessToast(updateDeadlinesType: DeadlineHandlerType,
+    selectedStudents: StudentExtensionTableColumnModel[],
+    selectedInstructors: InstructorExtensionTableColumnModel[]): void {
```
Let's make the following change to match the other functions, else it looks like the parameters are part of the method body.

```suggestion
  private showSucessToast(updateDeadlinesType: DeadlineHandlerType,
    selectedStudents: StudentExtensionTableColumnModel[],
    selectedInstructors: InstructorExtensionTableColumnModel[]
  ): void {
```

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.html:
```diff
@@ -0,0 +1,169 @@
+<fieldset [disabled]="isSubmitting">
```
I'm not sure if there's a better way, but I used fieldset so that I could disable all of the buttons at once instead of individually adding the disabled option.

samuelfangjw replied:
I guess it's fine, seems to be allowed by html spec.

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+        this.isLoadingAllInstructors = false;
+      }))
+      .subscribe(
+        (value: any[]) => {
```
Is it possible to make `value`'s type here `[Course, FeedbackSession]`? If so, we can simply destructure it instead of line 138 and line 140

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+    modalRef.componentInstance.feedbackSessionTimeZone = this.feedbackSessionTimeZone;
+
+    modalRef.componentInstance.onConfirmExtensionCallBack.subscribe((isNotifyDeadlines: boolean) => {
+      this.handleCreateAndDeleteDeadlines(selectedStudents, selectedInstructors, DeadlineHandlerType.CREATE,
```
nit: can line break before `DeadlineHandlerType.CREATE` to highlight the different param input from line 308

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+    modalRef.componentInstance.feedbackSessionTimeZone = this.feedbackSessionTimeZone;
+
+    modalRef.componentInstance.onConfirmExtensionCallBack.subscribe((isNotifyDeadlines: boolean) => {
+      this.handleCreateAndDeleteDeadlines(selectedStudents, selectedInstructors, DeadlineHandlerType.DELETE,
```
ditto

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+
+    if (updateDeadlinesType === DeadlineHandlerType.CREATE) {
+      selectedIndividuals.forEach((x) => {
+        record[x.email] = extensionTimestamp!;
```
In the spirit of defensive programming, what would happen if `extensionTimestamp` is `null`?

FergusMok replied:
Yeah, it's hard to enforce or prevent it to be null after merging them. Would it be better then to split the function back into separate functions?

moziliar replied:
Hmm that's a lot more code... I suppose this is fine as the wrong use would not make sense from the look of it (create deadline without timestamp)

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+  }
+
+  hasSelected(): boolean {
+    return [...this.studentsOfCourse, ...this.instructorsOfCourse].some((user) => user.isSelected);
```
Just being very pedantic: 
Will this be better?
```suggestion
    return this.studentsOfCourse.some((user) => user.isSelected)
        && this.instructorsOfCourse.some((user) => user.isSelected);
```

samuelfangjw replied:
Small mistake in the correction btw. This changes the logic, should be `||` not `&&`.

moziliar replied:
Right.. thanks

moziliar replied:
hence it is being pedantic: the original mandates both `student` and `instructor` to have `isSelected` which is a stronger coupling. Readability wise I feel the same way that the former is more readable.
@samuelfangjw what do you think? A tie-break from you?

samuelfangjw replied:
> the original mandates both student and instructor to have isSelected which is a stronger coupling.

Not sure if I understood this correctly, but I believe they are functionally equivalent? Both are checking if any student or instructor are selected.

No strong opinion on this, but prefer the correction because it better describes the intent in my opinion.

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.component.ts:
```diff
@@ -0,0 +1,528 @@
+   * Checks if at least one valid extension has been selected
+   */
+  hasSelectedValidForDeletion(): boolean {
+    return [...this.studentsOfCourse, ...this.instructorsOfCourse].some((user) => user.isSelected && user.hasExtension);
```
ditto

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.spec.ts:
```diff
@@ -0,0 +1,483 @@
+  const testStudent3: Student = {
+    email: 'alex@tmms.com',
+    courseId: 'testId',
+    name: 'alex',
```
```suggestion
    name: 'Alex',
```

moziliar commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/instructor-session-individual-extension-page.spec.ts:
```diff
@@ -0,0 +1,483 @@
+    jest.spyOn(feedbackSessionsService, 'getFeedbackSession').mockReturnValue(of(testFeedbackSession));
+
+    component.ngOnInit();
+    component.isLoadingAllInstructors = false;
+    component.isLoadingAllStudents = false;
+    component.isLoadingFeedbackSession = false;
```
Just checking: should these be `expect` instead?

moziliar commented:
Any updates to this PR? After resolving the conflict (might need to rebase), I will take a final scan. The logic should be good to go!

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,186 @@
+    return this.radioOption === RadioOptions.EXTEND_TO;
+  }
+
+  isCustomize(): boolean {
+    return this.isRadioExtendBy() && this.extendByDeadlineKey === 'Customize';
+  }
```
You mean after `isRadioExtendBy`? Yup makes sense

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-confirm-modal/individual-extension-confirm-modal.component.ts:
```diff
@@ -0,0 +1,138 @@
+
+  SortBy: typeof SortBy = SortBy;
+  SortOrder: typeof SortOrder = SortOrder;
+  sortStudentsBy: SortBy = SortBy.SESSION_END_DATE;
+  sortStudentOrder: SortOrder = SortOrder.DESC;
+  sortInstructorsBy: SortBy = SortBy.SESSION_END_DATE;
+  sortInstructorOrder: SortOrder = SortOrder.DESC;
```
I think putting all of the related variables into chunks would actually make it more readable, especially when there's many of them 

E.g from `individual-extension-page` 
![image](https://user-images.githubusercontent.com/10182564/162556225-96e7e1c1-9d92-4476-889b-fef9c444c0db.png)

moziliar replied:
I see. Minor nit coz when I read this block, it is a bit hard to discern what's going on there (the different length of `SortBy` and `SortOrder` makes the reading a bit jagged. Though I have no strong opinion about this. Can resolve this thread if it's ok with you:)

FergusMok commented on src/web/app/pages-instructor/instructor-session-individual-extension-page/individual-extension-date-modal/individual-extension-date-modal.component.ts:
```diff
@@ -0,0 +1,186 @@
+      return;
+    }
+
+    const extensionTimeString = this.dateDetailPipe.transform(
+      this.getExtensionTimestamp(),
+      this.feedbackSessionTimeZone,
+    );
+    const currentTimeString = this.dateDetailPipe.transform(
+      Date.now(),
+      this.feedbackSessionTimeZone,
+    );
```
Yup good idea, the pipe may actually not convey that it's adjusting the value

jayasting98 reviewed (approved):
Seems fine to me.

moziliar commented:
I'm largely good with this PR. Will take a final scan later today before approving!

samuelfangjw reviewed (approved):
Looks good to me!

moziliar reviewed (approved):
LGTM!

FergusMok commented:
Thanks everyone for taking the time to review this
